analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs

```diff
@@ -141,12 +160,12 @@ public ProgramState ResetStaticFieldConstraints(INamedTypeSymbol containingType)
 
         public override int GetHashCode() =>
             // VisitCount is not included, it's not part of Equals
-            HashCode.Combine(
-                HashCode.DictionaryContentHash(OperationValue),
-                HashCode.DictionaryContentHash(SymbolValue),
-                HashCode.DictionaryContentHash(CaptureOperation),
-                HashCode.EnumerableContentHash(PreservedSymbols),
-                HashCode.EnumerableContentHash(Exceptions));
+            hashCode ??= HashCode.Combine(
```

Comment:
Can you measure how many times we create `ProgramState` without ever calling getHashCode? In theory, we could do this math at creation time.. or at least inside the wrapped type for the sub-items.

Reply:
I did similar stats against Nancy project:
0x 145696
1x: 124010
2x: 13
3x: 18
3520x: 1 - And I confirmed it's the "Empty" one.

To me, it does not make sense to cache it. There's nothing to win, there's no cache-hit.

The only shortcut we can take is to `return 0` when reference equals to `Empty`.

Reply:
The saving is that in 124.010 cases we do calculate most likely a single hashcode for a single ImmutableDictionary instead of 5.

Reply:
Ok. If the main gain is on the partial sub-results, what about removing `hashCode` itself?

It's calculation should be rather cheap. And it's not reused much. It will also simplify the inits.

Reply:
As an alternative solution, we can make the hashcode implementation allocations free (my main concern here). I did so in #7012. There you can find an allocation compression between the base-line, this PR (caching), and an allocation-free implementation. I would recommend merging #7012 in any case and we may also merge this PR on top of it (which would additionally save some CPU cycles).

Reply:
Let's merge the other one. And measure the impact of this after.